### PR TITLE
[Login] Toggle password visibility (#5935)

### LINKED ISSUE #5935: [login / UserAccts] Toggle password visibility - Add clickable eye icon to un/mask input
Allow the user to unmask their password input by giving them a clickable icon in/near the password input box.  
Why? 
* increasingly common/standard in websites to help users spot their own typos when inputting complex passwords. 
* Avoids user lockout and saves both admin time and user frustration spent on password reset requests.

[Here's what it looks like](https://www.google.com/search?q=toggle+password+visibility&rlz=1C5CHFA_enCA859CA859&source=lnms&tbm=isch&sa=X&ved=2ahUKEwiDusXXqYjnAhWTBc0KHcWVD4kQ_AUoAXoECAwQAw&biw=1336&bih=647).  A grey version of this [eye-open glyphicon](https://www.w3schools.com/icons/tryit.asp?filename=trybs_ref_glyph_eye-open) is typically used.  
To re-mask (toggle on), use the icon of the eye with a line through it.  The password should never be unmasked by default.  Solution Examples (perhaps not directly applicable) available on [Stack Overflow](https://stackoverflow.com/questions/51121625/place-mask-unmask-button-inside-a-password-input-element) and [W3schools](https://www.w3schools.com/howto/howto_js_toggle_password.asp), etc.

(To verify) Should be applied to the following pages/modules: 
* Login
* Reset password (upon first login, e.g.)
* Request account
* User Accounts: Add/Edit User

### PR DESCRIPTION
**Pull request by laemtl**
## Brief summary of changes
I added the necessary script and CSS to add the password toggle visibility feature.
This is a first draft, with 2 versions.
- I have seen some suggestions to get rid of Jquery dependencies, so the first version is in plain javascript [a153a7d]. Let me know if I should migrate to a Jquery solution :)
- Also, as LORIS is migrating to React, a second version uses a React component [ac83ccf].

#### Testing instructions
1. On the Login page, click the eye icon on the password input field to toggle the visibility
2. On /user_accounts/my_preferences/, click the eye icon on the password input field to toggle the visibility
3. On /user_accounts/edit_user/$user, click the eye icon on the password input field to toggle the visibility

#### Link(s) to related issue(s)
* Resolves #5935

#### About me
GOSC applicant, happy to join the community!

@christinerogers

### CODE REVIEW

**Comment by johnsaigle** [comment]
This is great! We've been meaning to add this for a long time and it's awesome that you've completed it so quickly.

My JS skills are very weak so I can't approve this but the tests work great.

Screen Shot 2020-04-01 at 10 50 09
Screen Shot 2020-04-01 at 10 50 14

In addition to login and my_preferences, the `edit_user` also has a password field. Could you add the toggle there as well? You can find it under `https://$your-loris/user_accounts/edit_user/$user`

My understanding is that we don't want to use jQuery so ReactJS is the way to go. 👍

**Comment by maltheism** [comment]
@laemtl very cool feature! I've added the logic of this PR to one of my PRs and that will likely get merged later than your PR. See https://github.com/aces/Loris/pull/5828 for reference.

**Comment by laemtl** [comment]
@johnsaigle I added the component tio the edit_user page.

**Comment by maltheism** [comment]
Hi @laemtl, you may already know how to do the following but I thought to suggest because there are now conflicting files from latest merge that happened into master. You can rebase by first fetching aces `fetch aces` and then rebase `git rebase -i aces/master` while doing `git status` & `git rebase --continue` after fixing the conflicting files. Finally you can do `git push -f origin your_pr_branch_name` to force push the branch.

**Comment by laemtl** [comment]
@maltheism Rebase has been done! Thanks for notifying me.

**Comment by christinerogers** [comment]
This PR seems ready for final review on the `master` branch as soon as the 23 release is done. 

 @johnsaigle @maltheism  Could you officially Review when there's a free moment in mid-May? thanks

**Comment by pierre-p-s** [comment]
PR has conflicts (needs rebase).

I tested the edit user accounts, the preferences and the login, everything works.

**Comment by driusan** [comment]
I think the plain javascript version is definitely superior in this case, but can't be merged because your PR is currently set up for the React version.

The React version doesn't integrate well outside of a React context (ie in the smarty templates where the changes seem to all be..) or if javascript is disabled, while the plain javascript one degrates more gracefully to a normal HTML password input box if the javascript doesn't work for any reason.

**Comment by laemtl** [comment]
Sure, I will merge the JS version into the react one and adjust the pages changed accordingly,

**Comment by laemtl** [comment]
@driusan Change done.

**Comment by CamilleBeau** [comment]
Manual test went well! I thought the code looks good, but I'm still pretty new to js and react so may want a second reviewer.

File: htdocs/js/passwordVisibility.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+ * with a toggle option to display it in plain text
+ *
+ * For React integration, see React version (jsx/Password.js)
+ */
+
+/**
+ * Collect all input[type='password'] in the DOM
+ * and instantiate PwdInput
+ */
+class PasswordVisibility {
```
Comment (driusan) [inline htdocs/js/passwordVisibility.js:17]: This seems to be used as (and acts as) a function, there's no reason to make it a class.

Comment (laemtl) [reply htdocs/js/passwordVisibility.js]: @driusan  Fixed!

**Comment by driusan** [comment]
@laemtl this is failing Travis

**Comment by laemtl** [comment]
@driusan Solved.

**Comment by laemtl** [comment]
@maltheism, @driusan  Following the discussion of today's meeting, I think the js script needs to be migrated in jslib/.

File: package.json [hunk 2]
```diff
@@ -26,16 +26,17 @@
     "c3": "^0.7.15",
     "d3": "^5.15.0",
     "css-loader": "^3.4.2",
+    "alex": ">=8.0.1",
```
Comment (maltheism) [inline package.json:29]: Why is the `alex` package added?

**Comment by maltheism** [comment]
@laemtl I'm unsure what would be best. I did take and modify your code for a component I used in the reactification PR of the login module. I do think the way you have it in this PR is necessary for being used with smarty templates. We eventually want to get rid of the smarty templates completely. The Loris/jslib functions will be imported directly into react code to be used. I believe the code would then transpiled along with the react module code. The smarty template requires a <script> tag import I believe.

**Comment by laemtl** [comment]
So it's good to go as it is.

**Comment by driusan** [comment]
jslib doesn't exist yet, we can move it there later if need be.

**Pull request merged**